MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/utils/DemoAppHomeScreenAddressWidget.java

```diff
@@ -0,0 +1,230 @@
+ * the address in an Android home screen widget.
+ */
+public class DemoAppHomeScreenAddressWidget extends AppWidgetProvider implements
+  PermissionsListener {
+
+  private Context context;
+  private LocationChangeListeningActivityLocationCallback callback;
+  private static final long DEFAULT_INTERVAL_IN_MILLISECONDS = 1000L;
+  private static final long DEFAULT_MAX_WAIT_TIME = DEFAULT_INTERVAL_IN_MILLISECONDS * 5;
+
+  @Override
+  public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
+    this.context = context;
+    this.callback = new LocationChangeListeningActivityLocationCallback(context, appWidgetManager, appWidgetIds[0]);
+    checkPermissions(context);
+  }
+
+  @Override
+  public void onEnabled(Context context) {
+    // Add message/UI here if you'd like
+  }
+
+  @Override
+  public void onPermissionResult(boolean granted) {
+    if (granted) {
+      initializeLocationEngine(context);
+    } else {
+      Toast.makeText(context, R.string.user_location_permission_not_granted, Toast.LENGTH_SHORT).show();
+    }
+  }
+
+  /**
+   * Check whether location permissions are granted and then eventually initialize the LocationEngine
+   *
+   * @param context the application's context
+   */
+  private void checkPermissions(Context context) {
+    // Check if permissions are enabled and if not request
+    try {
+      if (PermissionsManager.areLocationPermissionsGranted(context)) {
+        // Create a location engine instance
+        initializeLocationEngine(context);
+      } else {
+        PermissionsManager permissionsManager = new PermissionsManager(this);
+        permissionsManager.requestLocationPermissions(getActivity(context));
+      }
+    } catch (NullPointerException nullPointerException) {
+      Toast.makeText(context, context.getString(R.string.enable_location), Toast.LENGTH_LONG).show();
+    }
+  }
+
+  /**
+   * Initialize the LocationEngine so that the device's current location can be retrieved
+   *
+   * @param context the application's context
+   */
+  @SuppressWarnings( {"MissingPermission"})
+  private void initializeLocationEngine(Context context) {
+    LocationEngine locationEngine = LocationEngineProvider.getBestLocationEngine(context);
+    LocationEngineRequest request = new LocationEngineRequest.Builder(DEFAULT_INTERVAL_IN_MILLISECONDS)
+      .setPriority(LocationEngineRequest.PRIORITY_HIGH_ACCURACY)
+      .setMaxWaitTime(DEFAULT_MAX_WAIT_TIME).build();
+
+    locationEngine.requestLocationUpdates(request, callback, getMainLooper());
+    locationEngine.getLastLocation(callback);
+  }
+
+  public Activity getActivity(Context context) {
+    if (context == null) {
+      return null;
+    } else if (context instanceof ContextWrapper) {
+      if (context instanceof Activity) {
+        return (Activity) context;
+      } else {
+        return getActivity(((ContextWrapper) context).getBaseContext());
+      }
+    }
+    return null;
+  }
```

Comment:
getActivity can be removed, only one if case should execute seeing the usages,  can be directly cast if needed